File: haskell-fuel-engine/src/FuelUpload/Summary.hs
module FuelUpload.Summary
  ( summarizeRows
  , outcomeFromRows
  ) where

import Data.List.NonEmpty (NonEmpty (..))
import FuelUpload.Domain.Decision
import FuelUpload.Domain.Primitive

summarizeRows :: [RowDecision] -> BatchSummary
summarizeRows =
  foldr addDecision emptySummary
  where
    emptySummary =
      BatchSummary
        { summaryAccepted = 0
        , summaryAcceptedWithWarnings = 0
        , summarySkippedDuplicates = 0
        , summaryRejected = 0
        , summaryFatal = 0
        , summaryTotalRows = 0
        }

    addDecision decision summary =
      case decision of
        Accepted _ ->
          countAccepted summary
        AcceptedWithWarnings _ _ ->
          countAcceptedWithWarnings summary
        SkippedDuplicate _ ->
          countSkipped summary
        Rejected _ ->
          countRejected summary
        Fatal _ ->
          countFatal summary

    countAccepted summary =
      bumpTotal summary
        { summaryAccepted = summaryAccepted summary + 1
        }

    countAcceptedWithWarnings summary =
      bumpTotal summary
        { summaryAccepted = summaryAccepted summary + 1
        , summaryAcceptedWithWarnings = summaryAcceptedWithWarnings summary + 1
        }

    countSkipped summary =
      bumpTotal summary
        { summarySkippedDuplicates = summarySkippedDuplicates summary + 1
        }

    countRejected summary =
      bumpTotal summary
        { summaryRejected = summaryRejected summary + 1
        }

    countFatal summary =
      bumpTotal summary
        { summaryFatal = summaryFatal summary + 1
        }

    bumpTotal summary =
      summary {summaryTotalRows = summaryTotalRows summary + 1}

outcomeFromRows :: [RowDecision] -> BatchOutcome
outcomeFromRows rows =
  case fatalErrors rows of
    firstFatal : remainingFatals ->
      BatchBlockedByFatal (firstFatal :| remainingFatals)
    [] ->
      BatchUploadable

fatalErrors :: [RowDecision] -> [FatalError]
fatalErrors =
  foldr collectFatal []
  where
    collectFatal decision errors =
      case decision of
        Fatal fatalError -> fatalError : errors
        Accepted _ -> errors
        AcceptedWithWarnings _ _ -> errors
        SkippedDuplicate _ -> errors
        Rejected _ -> errors

File: rust-fuel-engine/src/engine/summary.rs
use crate::domain::*;

impl Summary {
    pub fn from_rows(rows: &[RowDecision]) -> Self {
        rows.iter().fold(
            Self {
                total_rows: rows.len(),
                ..Self::default()
            },
            |mut summary, decision| {
                match decision {
                    RowDecision::Accepted(_) => {
                        summary.accepted += 1;
                    }
                    RowDecision::Warning { warnings, .. } => {
                        summary.accepted += 1;
                        summary.accepted_with_warnings += 1;
                        summary.warnings += warnings.len();
                    }
                    RowDecision::SkippedDuplicate(_) => {
                        summary.skipped_duplicates += 1;
                    }
                    RowDecision::Rejected(_) => {
                        summary.rejected += 1;
                    }
                    RowDecision::Fatal(_) => {
                        summary.fatal_errors += 1;
                    }
                }
                summary
            },
        )
    }
}
